WcaOnRails/app/models/result.rb

```diff
@@ -1,16 +1,105 @@
+  belongs_to :country, foreign_key: :countryId
   belongs_to :person, -> { current }, primary_key: :wca_id, foreign_key: :personId
   belongs_to :round, foreign_key: :roundId
   belongs_to :event, foreign_key: :eventId
+  belongs_to :format, foreign_key: :formatId
 
   scope :podium, -> { joins(:round).merge(Round.final_rounds).where(pos: [1..3]).where("best > 0") }
   scope :winners, -> { joins(:round, :event).merge(Round.final_rounds).where("pos = 1 and best > 0").order("Events.rank") }
 
+  validate :validate_each_solve
+  def validate_each_solve
+    solve_times.each_with_index do |solve_time, i|
+      unless solve_time.valid?
+        errors.add(:"value#{i + 1}", solve_time.errors.messages[:base].join(" "))
+      end
+    end
+  end
+
+  validate :validate_number_of_solves
+  def validate_number_of_solves
+    return errors.add(:competitionId, "invalid") unless competition
+    return errors.add(:countryId, "invalid") unless country
+    return errors.add(:roundId, "invalid") unless round
+    return errors.add(:eventId, "invalid") unless event
+    return errors.add(:formatId, "invalid") unless format
+    return errors.add(:base, invalid_number_of_solves_reason) if invalid_number_of_solves_reason
+  end
+
+  def invalid_number_of_solves_reason
+    return "Invalid format" unless format
+    return "Invalid round" unless round
+    return "Cannot skip all solves." if solve_times.all?(&:skipped?)
+
+    unless solve_times.drop_while(&:unskipped?).all?(&:skipped?)
+      return "Skipped solves must all come at the end."
+    end
+
+    unskipped_count = solve_times.length - solve_times.count(&:skipped?)
+    if round.combined?
+      if unskipped_count > format.expected_solve_count
+        return "Expected at most #{hlp.pluralize(format.expected_solve_count, 'solve')}, but found #{unskipped_count}."
+      end
+    else
+      if unskipped_count != format.expected_solve_count
+        return "Expected #{hlp.pluralize(format.expected_solve_count, 'solve')}, but found #{unskipped_count}."
+      end
+    end
+  end
+
+  def average_is_not_computable_reason
+    # To compute the average, we need to have a valid number of solves,
+    # and we need to know what event we are dealing with (because
+    # 333fm is computed differently than other events).
+    !event ? "Event needed to compute average" : invalid_number_of_solves_reason
+  end
+
+  validate :validate_average
+  def validate_average
+    return if average_is_not_computable_reason
+
+    correct_average = compute_correct_average
+    errors.add(:average, "should be #{correct_average}") if correct_average != average
+  end
+
+  validate :validate_best
+  def validate_best
+    correct_best = compute_correct_best
+    errors.add(:best, "should be #{correct_best}") if correct_best != best
+  end
+
+  def compute_correct_best
+    sorted_solves.first&.wca_value
+  end
+
+  def compute_correct_average
+    fail average_is_not_computable_reason if average_is_not_computable_reason
```

Comment:
This is actually an assertion. Someone calling `compute_correct_average` is expected to first check if the average is actually computable.

I'm definitely open to restructuring this. Are you free to chat today?